scripts/run_e2e_tests.py

```diff
@@ -559,10 +633,80 @@ def main(args=None):
+                plaintext = simplecrypt.decrypt(
+                    google_auth_decode_password, ciphertext).decode('utf-8')
+                dec_file.write(plaintext)
+
+        sheets_scopes = ['https://www.googleapis.com/auth/spreadsheets']
+        creds = service_account.Credentials.from_service_account_file(
+            'auth.json', scopes=sheets_scopes)
+        sheet = googleapiclient.discovery.build(
+            'sheets', 'v4', credentials=creds).spreadsheets()
+        flaky_tests_list = get_flaky_tests_data_from_sheets(sheet)
+
+    suite_name = parsed_args.suite.lower()
+    if len(flaky_tests_list) > 0 and p.returncode != 0:
+        for i, line in enumerate(output_lines):
+            if line == '*                    Failures                    *':
+                test_name = output_lines[i + 3][3:].strip().lower()
+
+                # Remove coloring characters.
+                ansi_escape = re.compile(
+                    r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])')
+                failure_log = ansi_escape.sub('', output_lines[i + 4])
+                failure_log = failure_log.strip().lower()
+                for index, row in enumerate(flaky_tests_list):
+                    flaky_suite_name = row[0].strip().lower()
+                    flaky_test_message = row[1].strip().lower()
+                    flaky_error_message = row[2].strip().lower()
+                    if (
+                            suite_name == flaky_suite_name or
+                            flaky_suite_name == '[general]'):
+                        if (
+                                test_name == flaky_test_message or
+                                flaky_test_message == 'many'):
+                            if flaky_error_message in failure_log:
+                                update_flaky_tests_count(sheet, index, row[3])
+                                try:
+                                    cleanup_portserver(portserver_process)
+                                    cleanup()
+                                except Exception: # pragma: no cover
+                                    # This is marked as no cover because the
+                                    # exception happens due to some processes
+                                    # running on the local system, which might
+                                    # interfere with the cleanup stuff. This is
+                                    # added as a failsafe to make sure that
+                                    # even when it throws an exception, the
+                                    # test is retried.
+                                    pass # pragma: no cover
+                                return 'flake'
     sys.exit(p.returncode)
 
 
+def main(args=None):
+    """Run tests, rerunning at most MAXIMUM_RUNS times if they flake."""
+    for _ in python_utils.RANGE(MAXIMUM_RUNS):
+        flake_state = run_tests(args=args)
+        if flake_state != 'flake':
```

Comment:
Should we structure it this way:

```
get_flkay_test_data_sheet:
    data = get_flkay_test_data_sheet()
    return get_flaky_tests_data_from_sheets(data)

is_test_output_flaky(output):
    logs = get_flkay_test_logs()
    for log in logs:
         if is_output_similar_to_log(output, log)
             return True
    return False

main:
    return_code = 1
    Loop max_tries:
         output, return_code = run_test()
         if is_test_output_faky(output):
             update_falkey_test_count() [OR do it this inside the is_test_output_flkay() call]
             print 'The test result looks flaky, rerunning the tests (%<attempt number>)'
         else:
             break
 
    sys.exist(return_code)
```

[This is just to make the code structured and readable!]